Added Meta, Charset, and Transitional parameters to ConvertTo-HTML

ergo3114 opened the pull request:
Related #3267

Added the Meta, Chartset, and Transitional parameters to ConvertTo-HTML

**Meta** Allows the user to insert `<meta>` tag information such as refresh, author, and keywords
**Charset** Sets the html encoding of the page, denoted by the `<charset>` html tag
**Transitional** A switch parameter that allows for the insertion of XHTML Transitional DTD DOCTYPE, the default being XHTML Strict DTD

Edit: Documentation can be found [here](https://github.com/PowerShell/PowerShell-Docs/pull/1432)

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to CONTRIBUTING.md.

-->

ergo3114 committed:
- Added functionality to convert-HTML.cs to include a charset parameter

ergo3114 committed:
- Fixed issues on the charset parameter where flag was always set to false

ergo3114 committed:
- Moved Meta tag into Head tag

ergo3114 committed:
- Took away charset defaullt value

ergo3114 committed:
- Added Transitional switch to allow users to switch between strict and transitional DOCTYPEs
- Updated Meta parameter to handle http-equiv properties

msftclas commented:
@ergo3114,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

ergo3114 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +437,33 @@ protected override void BeginProcessing()
                 WriteObject("<head>");
+                if(_charsetSpecified)
+                {
+                    WriteObject("<meta charset=\"" + _charset + "\">");
```
Validation would be great. I don't see too many changes in the charset list but it would be great if the validation were dynamic pulling from w3schools or something. I was trying to find documentation on what the browser does when a non standard charset is specified; such as "foo". I believe it analyzes the html and determines the best or just defaults to UTF-8 for HTML5.

ergo3114 replied:
I will put something in and resubmit it.

ergo3114 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +437,33 @@ protected override void BeginProcessing()
+                            WriteObject("<meta name=\"" + s + "\" content=\"" + _meta[s] + "\">");
+                        }
+                        else {
+                            WriteObject("<meta http-equiv=\"" + s + "\" content=\"" + _meta[s] + "\">");
```
It looks like you can validate the meta tag. [w3schools](https://www.w3schools.com/tags/tag_meta.asp) shows the schema for the meta tag.

ergo3114 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +437,33 @@ protected override void BeginProcessing()
+                    {
+                        if(s != "content-type" && s != "default-style" && s !="refresh")
+                        {
+                            WriteObject("<meta name=\"" + s + "\" content=\"" + _meta[s] + "\">");
```
What would encoding gain us here? I am not familiar with that.

ergo3114 committed:
- Added validation to the Charset parameter
- Set the output of Charset to upper case if a lowercase value is given

ergo3114 committed:
- Put the Meta parameter output inside the HTML head tag

ergo3114 committed:
- Added filtering for the Meta parameter to only allow defined HTML meta values

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,68 @@ public string[] PreContent
+            set
+            {
+                _meta = value;
+                _metaSpecified = true;
```
I believe the typical pattern is to use [InvocationInfo.BoundParameters](https://msdn.microsoft.com/en-us/library/system.management.automation.invocationinfo.boundparameters(v=vs.85).aspx)

ergo3114 replied:
I was unable to get InvocationInfo.BoundParameters to work. I read over InvocationInfo.cs and ImplicitRemotingCommands.cs to see if I can make sense of it but I am at a lose here.

iSazonov replied:
We can use an auto property and `Meta == null` to check if user assign a value.

SteveL-MSFT replied:
That should be fine, too

iSazonov replied:
@ergo3114 Please address this.

ergo3114 replied:
I see this as two ways to do the same thing. What are you looking to gain from setting a default value versus checking a flag? Just curious.

SteveL-MSFT replied:
A quick search shows we already have inconsistency so I'm fine staying with this pattern

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +438,33 @@ protected override void BeginProcessing()
+                {
+                    foreach(string s in _meta.Keys)
+                    {
+                        if(s == "content-type" && s == "default-style" && s =="refresh")
```
Shouldn't this be `||`?  In any case, there's more `http-equiv` than this list.  Perhaps we should just have a separate parameter for `http-equiv` meta tags and let the user put whatever they want in it?  Also, I believe the names are case-insensitive so `Content-Type` is valid.

ergo3114 replied:
This has been changed.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,68 @@ public string[] PreContent
+        /// </summary>
+        [Parameter(ParameterSetName = "Page")]
+        [ValidateNotNullOrEmpty]
+        [ValidateSet("UTF-8", "ISO-8859-1","ASCII","ANSI",IgnoreCase=true)]
```
Technically, the allowed set of (legacy) encodings is quite large although HTML5 I believe has standardized on UTF-8.  So perhaps this shouldn't have a `ValidateSet`?

ergo3114 replied:
I have removed to the ValidateSet.

SteveL-MSFT commented:
Please add some tests

SteveL-MSFT reviewed (changes requested):
Please add some tests

ergo3114 committed:
- Changed the condition for http-equiv from && to ||
- Removed Validateset from Charset parameter to allow for legacy character sets
- Added test for Meta, Charset, and Transitional property on ConvertTo-HTML command

ergo3114 committed:
- Removed static declaration while attempting to using InvocationInfo.BoundParameters

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,67 @@ public string[] PreContent
+            set
+            {
+                _charset = value.ToUpper();
+                _charsetSpecified = true;
```
Why do we convert a charset to upper case?
It seems we can remove this and use auto property.

iSazonov replied:
Closed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,67 @@ public string[] PreContent
+            set
+            {
+                _transitional = value;
+            }
```
We can use an auto property for a switch parameter without getter/setter.

iSazonov replied:
Closed.

ergo3114 committed:
- Removed defined get/set functions from the transitional switch parameter
- Removed private variable for transitional switch parameter
- Removed toUpper() functin from charset

ergo3114 committed:
- Removed commented out testing

ergo3114 committed:
Merge remote-tracking branch 'refs/remotes/PowerShell/master'

ergo3114 committed:
- Removed \1 in the root folder

ergo3114 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,56 @@ public string[] PreContent
+        /// </summary>
+        [Parameter(ParameterSetName = "Page")]
+        [ValidateNotNullOrEmpty]
+        public string Charset
```
Certainly can. Are you concerned with someone putting HTML into the parameter and injecting malicious code?

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +426,33 @@ protected override void BeginProcessing()
+                        else if (s == "application-name" || s == "author" || s == "description" || s == "generator" || s == "keywords" || s == "viewport") {
+                            WriteObject("<meta name=\"" + s + "\" content=\"" + _meta[s] + "\">");
+                        }
+                    }
```
Perhaps we should output a warning for any other meta that we didn't include in case the user made a typo

ergo3114 replied:
I put in an error and have pushed to my repo. If you have an example of a warning I will gladly change it to a warning.

SteveL-MSFT replied:
Rather than doing the validation above and going through the hashtable twice, I think it might be better to change these lines to a switch statement and have it default to an error

ergo3114 committed:
- Added pattern validation to the charset parameter
- Added access to the Meta, Charset, and Transitional parameters while in a workflow

ergo3114 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +426,33 @@ protected override void BeginProcessing()
+                }
+                if(_metaSpecified)
+                {
+                    foreach(string s in _meta.Keys)
```
What do you mean?

ergo3114 replied:
I see what you mean and have changed my code to reflect that.

ergo3114 committed:
- Added error message in resource file
- Added error when meta tag could not be parsed (should be warning and continue)

ergo3114 commented:
Documentation updated and added per [PowerShell-Docs #1432](https://github.com/PowerShell/PowerShell-Docs/pull/1432)

ergo3114 committed:
- Fixed error introduced in test on push 0f9279b when testing meta parameter with invalid properties

ergo3114 committed:
- Removed refresh as a valid meta property

ergo3114 committed:
- Updated Meta parameter validation to a switch statement from an if statement
- Updated error message on Meta parameter validation and removed refresh Reference: 06e9751

ergo3114 commented:
Documentation for this PR was merged July 31th. [1432](https://github.com/PowerShell/PowerShell-Docs/pull/1432)

ergo3114 committed:
- Removed comments

ergo3114 commented:
Removed comments. Keep me posted with anything else you find and I'll be glad to work on your request.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -376,9 +427,44 @@ protected override void BeginProcessing()
+                {
+                    foreach(string s in _meta.Keys)
+                    {
+                        switch(s){
```
Based on what I could find, meta tag names are not case sensitive (although best practice is to use lowercase), so it may be good to throw in a `ToLower()`, but you're using a hashtable which is case sensitive so the same the same tag could show up twice.  Perhaps add code to check if a particular tag is already used and return an error.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Html.Tests.ps1:
```diff
@@ -116,5 +116,65 @@ After the object
+    }
+
+    It "Test ConvertTo-HTML meta with invalid properties should throw"{
+        { ($customObject | ConvertTo-HTML -Meta @{"authors"="John Doe";"keyword"="PowerShell,PSv6"}) -join $newLine } | Should Throw
```
it would be better to check for a specific errorid if one is expected (and consistent)

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Html.Tests.ps1:
```diff
@@ -116,5 +116,65 @@ After the object
+    }
+
+    It "Test ConvertTo-HTML transitional"{
+        $returnString = ($customObject | ConvertTo-HTML -Transitional) -join $newLine
```
do you want to validate the entire document or just the first line? If the later, would:
```
$result = $customObject | ConvertTo-HTML -Transitional | Select-Object -First 1
$result | should be '<!DOCTYPE html PUBLIC...
```
be sufficient?

ergo3114 commented:
My development computer crashed and I am having issues getting things set back up. I will continue updating this PR when I fix my issues.

ergo3114 commented:
I believe I am back in the game with continuing this PR. I will work on items that were left unattended in the coming days.

ergo3114 committed:
- Added X-UA-Compatible as an approved meta item
- Set meta hashtable keys to lower before checking in switch statement
- Updated test for the transitional parameter to check only the first line of the result
- Updated test to be more specific on the error message that is returned when thrown
- Added logic to prevent duplicate meta tags

ergo3114 committed:
- Pull from upstream

ergo3114 committed:
- Updated Test ConvertTo-HTML meta with invalid properties should thow

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -228,6 +228,57 @@ public string[] PreContent
+        /// <returns></returns>
+        [Parameter(ParameterSetName = "Page")]
+        [ValidateNotNullOrEmpty]
+        public SwitchParameter Transitional;
```
I think you want { get; set; } here instead of a field.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -364,9 +415,49 @@ protected override void BeginProcessing()
+                                    break;
+                            }
+                            useditems.Add(s);
+                        }
```
seems like we should give a warning if we are dropping duplicates

ergo3114 replied:
I have changed it to a warning and decided to keep the meta tags that don't match the switch.

ergo3114 committed:
- added {get;set;} for the Transitional vaiable

ergo3114 committed:
- Added warning message to meta tag from error

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/convert-HTML.cs:
```diff
@@ -461,8 +461,19 @@ protected override void BeginProcessing()
-                                    ThrowTerminatingError (new ErrorRecord(exc, "MetaPropertyNotFound", ErrorCategory.ParserError, null));
+                                    //Exception exc = new NotSupportedException(StringUtil.Format(ConvertHTMLStrings.MetaPropertyNotFound, s));
+                                    MshCommandRuntime mshCommandRuntime = this.CommandRuntime as MshCommandRuntime;
+                                    string Message = "Accepted meta properties are content-type, default-style, application-name, author, description, generator, keywords, x-ua-compatible, and viewport. The meta pair: " + s + " and " + _meta[s] + " may not function correctly.";
```
Add a new string in the resx file so that it can be localized

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Html.Tests.ps1:
```diff
@@ -136,8 +136,9 @@ After the object
-    It "Test ConvertTo-HTML meta with invalid properties should throw"{
-        { ($customObject | ConvertTo-HTML -Meta @{"authors"="John Doe";"keywords"="PowerShell,PSv6"}) -join $newLine } | Should Throw "authors is not a supported meta property. Accepted meta properties are content-type, default-style, application-name, author, description, generator, keywords, x-ua-compatible, and viewport."
+    It "Test ConvertTo-HTML meta with invalid properties should throw warning" {
+        ($customObject | ConvertTo-HTML -Meta @{"authors"="John Doe";"keywords"="PowerShell,PSv6"} 3>&1) -match "Accepted meta properties are" | Should Be $true
```
Since the message could be localized, you probably just need to check that the warning contains the Meta tags you used in the test

ergo3114 committed:
- Pull warning from resource file

ergo3114 committed:
- Removed commented out code

ergo3114 committed:
- Repush for Appveyor test

ergo3114 committed:
- Repushing after rebase

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Html.Tests.ps1:
```diff
@@ -116,5 +116,52 @@ After the object
+    }
+
+    It "Test ConvertTo-HTML meta with invalid properties should throw warning" {
+        ($customObject | ConvertTo-HTML -Meta @{"authors"="John Doe";"keywords"="PowerShell,PSv6"} 3>&1) -match "Accepted meta properties are" | Should Be $true
```
Since the warning message can be localized, you shouldn't check against the English string.  Instead, check that the warning message contains the meta pair.

ergo3114 committed:
- Changed to pull warning message from resouce file

ergo3114 committed:
- Changed test for invalid meta keys to look for key pair in verbose output
